subsys/fs/nvs/nvs.c

```diff
@@ -574,17 +587,42 @@ static int nvs_startup(struct nvs_fs *fs)
 		}
 	}
 
+	u32_t data_wra_offs = 0;
 	/* addr contains address of the last ate in the most recent sector
 	 * search for the first ate containing all 0xff
 	 */
 	while (1) {
+		if ((addr & ADDR_OFFS_MASK) < ate_size) {
+			/* return -EDEADLK; */
```

Comment:
Why did you put this comment here if we are only breaking out of the loop and not returning an error?

Reply:
Initially I had considered this as an error in storage, but then I found out that this can be proper state as well. This commented line will be removed.

Reply:
@nvlsianpu, it cannot be a proper state. There should always be one empty ate in a sector. Otherwise there would be no fallback option when the sector end ate write would fail. NVS considers each ate to be valid (but it will not use the ate when the crc8 is invalid). If there is no empty ate you will always reach the 0 address.

Reply:
@nvlsianpu, @kapi-no, something that is wrong with nvs is that a fallback method is needed for the end sector ate. This could be avoided by not writing a sector end ate but by writing the needed information to the start of the next sector, in that case it could be checked at startup (and/or after the ate write) if the ate is correct and if needed erased and rewritten. Although it is not so straight forward (a combination of the ate write address, crc8 and data offset should be used to determine the last valid ate), it is possible.